File: sdk/rust/src/model_variant.rs
//! A single model variant backed by [`ModelInfo`].

use std::fmt;
use std::path::PathBuf;
use std::sync::Arc;

use serde_json::json;

use crate::catalog::CacheInvalidator;
use crate::detail::core_interop::CoreInterop;
use crate::detail::ModelLoadManager;
use crate::error::Result;
use crate::openai::AudioClient;
use crate::openai::ChatClient;
use crate::types::ModelInfo;

/// Represents one specific variant of a model (a particular id within an alias
/// group).
#[derive(Clone)]
pub struct ModelVariant {
    info: ModelInfo,
    core: Arc<CoreInterop>,
    model_load_manager: Arc<ModelLoadManager>,
    cache_invalidator: CacheInvalidator,
}

impl fmt::Debug for ModelVariant {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("ModelVariant")
            .field("id", &self.id())
            .field("alias", &self.alias())
            .finish()
    }
}

impl ModelVariant {
    pub(crate) fn new(
        info: ModelInfo,
        core: Arc<CoreInterop>,
        model_load_manager: Arc<ModelLoadManager>,
        cache_invalidator: CacheInvalidator,
    ) -> Self {
        Self {
            info,
            core,
            model_load_manager,
            cache_invalidator,
        }
    }

    /// The full [`ModelInfo`] metadata for this variant.
    pub fn info(&self) -> &ModelInfo {
        &self.info
    }

    /// Unique identifier.
    pub fn id(&self) -> &str {
        &self.info.id
    }

    /// Alias shared with sibling variants.
    pub fn alias(&self) -> &str {
        &self.info.alias
    }

    /// Check whether the variant is cached locally by querying the native
    /// core.
    ///
    /// Each call performs a full IPC round-trip. When checking many variants,
    /// prefer [`Catalog::get_cached_models`] which fetches the full list in a
    /// single call.
    pub async fn is_cached(&self) -> Result<bool> {
        let raw = self
            .core
            .execute_command_async("get_cached_models".into(), None)
            .await?;
        if raw.trim().is_empty() {
            return Ok(false);
        }
        let cached_ids: Vec<String> = serde_json::from_str(&raw)?;
        Ok(cached_ids.iter().any(|id| id == &self.info.id))
    }

    /// Check whether the variant is currently loaded into memory.
    pub async fn is_loaded(&self) -> Result<bool> {
        let loaded = self.model_load_manager.list_loaded().await?;
        Ok(loaded.iter().any(|id| id == &self.info.id))
    }

    /// Download the model variant. If `progress` is provided, it receives
    /// human-readable progress strings as the download proceeds.
    pub async fn download<F>(&self, progress: Option<F>) -> Result<()>
    where
        F: FnMut(&str) + Send + 'static,
    {
        let model_param = if self.info.provider_type.eq_ignore_ascii_case("huggingface") {
            &self.info.uri
        } else {
            &self.info.id
        };
        let params = json!({ "Params": { "Model": model_param } });
        match progress {
            Some(cb) => {
                self.core
                    .execute_command_streaming_async("download_model".into(), Some(params), cb)
                    .await?;
            }
            None => {
                self.core
                    .execute_command_async("download_model".into(), Some(params))
                    .await?;
            }
        }
        self.cache_invalidator.invalidate();
        Ok(())
    }

    /// Return the local file-system path where this variant is stored.
    pub async fn path(&self) -> Result<PathBuf> {
        let params = json!({ "Params": { "Model": self.info.id } });
        let path_str = self
            .core
            .execute_command_async("get_model_path".into(), Some(params))
            .await?;
        Ok(PathBuf::from(path_str))
    }

    /// Load the variant into memory.
    pub async fn load(&self) -> Result<()> {
        self.model_load_manager.load(&self.info.id).await
    }

    /// Unload the variant from memory.
    pub async fn unload(&self) -> Result<String> {
        self.model_load_manager.unload(&self.info.id).await
    }

    /// Remove the variant from the local cache.
    pub async fn remove_from_cache(&self) -> Result<String> {
        let params = json!({ "Params": { "Model": self.info.id } });
        let result = self
            .core
            .execute_command_async("remove_cached_model".into(), Some(params))
            .await?;
        self.cache_invalidator.invalidate();
        Ok(result)
    }

    /// Create a [`ChatClient`] bound to this variant.
    pub fn create_chat_client(&self) -> ChatClient {
        ChatClient::new(self.info.id.clone(), Arc::clone(&self.core))
    }

    /// Create an [`AudioClient`] bound to this variant.
    pub fn create_audio_client(&self) -> AudioClient {
        AudioClient::new(self.info.id.clone(), Arc::clone(&self.core))
    }
}

File: sdk/js/src/modelVariant.ts
import { CoreInterop } from './detail/coreInterop.js';
import { ModelLoadManager } from './detail/modelLoadManager.js';
import { ModelInfo } from './types.js';
import { ChatClient } from './openai/chatClient.js';
import { AudioClient } from './openai/audioClient.js';
import { ResponsesClient } from './openai/responsesClient.js';
import { IModel } from './imodel.js';

/**
 * Represents a specific variant of a model (e.g., a specific quantization or format).
 * Contains the low-level implementation for interacting with the model.
 */
export class ModelVariant implements IModel {
    private _modelInfo: ModelInfo;
    private coreInterop: CoreInterop;
    private modelLoadManager: ModelLoadManager;

    constructor(modelInfo: ModelInfo, coreInterop: CoreInterop, modelLoadManager: ModelLoadManager) {
        this._modelInfo = modelInfo;
        this.coreInterop = coreInterop;
        this.modelLoadManager = modelLoadManager;
    }

    /**
     * Gets the unique identifier of the model variant.
     * @returns The model ID.
     */
    public get id(): string {
        return this._modelInfo.id;
    }

    /**
     * Gets the alias of the model.
     * @returns The model alias.
     */
    public get alias(): string {
        return this._modelInfo.alias;
    }

    /**
     * Gets the detailed information about the model variant.
     * @returns The ModelInfo object.
     */
    public get modelInfo(): ModelInfo {
        return this._modelInfo;
    }

    /**
     * Checks if the model variant is cached locally.
     * @returns True if cached, false otherwise.
     */
    public get isCached(): boolean {
        const cachedModels: string[] = JSON.parse(this.coreInterop.executeCommand("get_cached_models"));
        return cachedModels.includes(this._modelInfo.id);
    }

    /**
     * Checks if the model variant is loaded in memory.
     * @returns True if loaded, false otherwise.
     */
    public async isLoaded(): Promise<boolean> {
        const loadedModels = await this.modelLoadManager.listLoaded();
        return loadedModels.includes(this._modelInfo.id);
    }

    /**
     * Downloads the model variant.
     * @param progressCallback - Optional callback to report download progress.
     *   Receives the status name (e.g. "Fetching model", "Downloading model") and
     *   a progress value. During the "Fetching model" phase the value is a heartbeat
     *   tick; during "Downloading model" it is the percentage (0-100).
     */
    public async download(progressCallback?: (name: string | null, progress: number) => void): Promise<void> {
        const request = { Params: { Model: this._modelInfo.id } };
        if (!progressCallback) {
            this.coreInterop.executeCommand("download_model", request);
        } else {
            await this.coreInterop.executeCommandStreaming("download_model", request, (chunk: string) => {
                const sepIndex = chunk.indexOf('|');
                if (sepIndex >= 0) {
                    const name = chunk.substring(0, sepIndex) || null;
                    const percent = parseFloat(chunk.substring(sepIndex + 1)) || 0;
                    progressCallback(name, percent);
                } else {
                    // Backwards compatibility: plain numeric progress
                    const progress = parseFloat(chunk);
                    if (!isNaN(progress)) {
                        progressCallback(null, progress);
                    }
                }
            });
        }
    }

    /**
     * Gets the local file path of the model variant.
     * @returns The local file path.
     */
    public get path(): string {
        const request = { Params: { Model: this._modelInfo.id } };
        return this.coreInterop.executeCommand("get_model_path", request);
    }

    /**
     * Loads the model variant into memory.
     * @returns A promise that resolves when the model is loaded.
     */
    public async load(): Promise<void> {
        await this.modelLoadManager.load(this._modelInfo.id);
    }

    /**
     * Removes the model variant from the local cache.
     */
    public removeFromCache(): void {
        this.coreInterop.executeCommand("remove_cached_model", { Params: { Model: this._modelInfo.id } });
    }

    /**
     * Unloads the model variant from memory.
     * @returns A promise that resolves when the model is unloaded.
     */
    public async unload(): Promise<void> {
        await this.modelLoadManager.unload(this._modelInfo.id);
    }

    /**
     * Creates a ChatClient for interacting with the model via chat completions.
     * @returns A ChatClient instance.
     */
    public createChatClient(): ChatClient {
        return new ChatClient(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates an AudioClient for interacting with the model via audio operations.
     * @returns An AudioClient instance.
     */
    public createAudioClient(): AudioClient {
        return new AudioClient(this._modelInfo.id, this.coreInterop);
    }

    /**
     * Creates a ResponsesClient for interacting with the model via the Responses API.
     * @param baseUrl - The base URL of the Foundry Local web service.
     * @returns A ResponsesClient instance.
     */
    public createResponsesClient(baseUrl: string): ResponsesClient {
        return new ResponsesClient(baseUrl, this._modelInfo.id);
    }
}
